// @parcel/core/lib/public/utils.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getInternalAsset = getInternalAsset;
exports.assetGraphVisitorToInternal = assetGraphVisitorToInternal;
exports.getBundleGroupId = void 0;

var _Asset = require("./Asset");

var _assert = _interopRequireDefault(require("assert"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const getBundleGroupId = bundleGroup => 'bundle_group:' + bundleGroup.entryAssetId;

exports.getBundleGroupId = getBundleGroupId;

function getInternalAsset(assetGraph, publicAsset) {
  let node = assetGraph.getNode(publicAsset.id);
  (0, _assert.default)(node != null && (node.type === 'asset' || node.type === 'asset_reference'));
  return node.value;
}

function assetGraphVisitorToInternal(visit) {
  if (typeof visit === 'function') {
    return assetGraphTraversalToInternal(visit);
  }

  return {
    enter: visit.enter ? assetGraphTraversalToInternal(visit.enter) : undefined,
    exit: visit.exit ? assetGraphTraversalToInternal(visit.exit) : undefined
  };
}

function assetGraphTraversalToInternal(visit) {
  return (asset, context, actions) => {
    return visit(new _Asset.Asset(asset), context, actions);
  };
}
