// @parcel/reporter-cli/lib/Spinner.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _cliSpinners = _interopRequireDefault(require("cli-spinners"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

class Spinner extends _react.default.Component {
  constructor(...args) {
    super(...args);

    _defineProperty(this, "timer", void 0);

    _defineProperty(this, "state", {
      frame: 0
    });

    _defineProperty(this, "switchFrame", () => {
      const {
        frame
      } = this.state;
      const spinner = this.getSpinner();
      const isLastFrame = frame === spinner.frames.length - 1;
      const nextFrame = isLastFrame ? 0 : frame + 1;
      this.setState({
        frame: nextFrame
      });
    });
  }

  getSpinner() {
    return _cliSpinners.default[this.props.type] || _cliSpinners.default.dots;
  }

  render() {
    const spinner = this.getSpinner();
    return spinner.frames[this.state.frame];
  }

  componentDidMount() {
    const spinner = this.getSpinner();
    this.timer = setInterval(this.switchFrame, spinner.interval);
  }

  componentWillUnmount() {
    clearInterval(this.timer);
  }

}

exports.default = Spinner;

_defineProperty(Spinner, "defaultProps", {
  type: 'dots'
});
